### PR TITLE
Validate possible operations for Grid suggestion

### PR DESCRIPTION
**Pull request by andreyvelich**
Fixes: https://github.com/kubeflow/katib/issues/1122.

I added validation for Chocolate Grid suggestion to check if `max_trial_count` >= all possible search space combination.
Also, I moved `call_validate()` test function to utils to use it in Chocolate also.

/assign @gaocegege @johnugeorge 
/cc @sperlingxx

### CODE REVIEW

**Comment by elikatsis** [comment]
Hello @andreyvelich,

Me and @StefanoFioravanzo have been tinkering with Katib for some time now and have a different proposal on tackling #1122. We will describe it in detail as a comment to that issue by the end of the week.

Would you mind holding this PR until then and decide afterwards whether to merge this or not?

Thanks in advance!

**Comment by andreyvelich** [comment]
Hi @elikatsis, sure.
Is this different from regular Validation Algorithm Settings?

File: pkg/suggestion/v1beta1/chocolate/service.py [hunk 1]
```diff
@@ -4,12 +4,15 @@
 from pkg.apis.manager.v1beta1.python import api_pb2
 from pkg.apis.manager.v1beta1.python import api_pb2_grpc
 
-from pkg.suggestion.v1beta1.internal.constant import DOUBLE
+from pkg.suggestion.v1beta1.internal.constant import (INTEGER, DOUBLE, CATEGORICAL, DISCRETE)
```
Comment (gaocegege) [inline pkg/suggestion/v1beta1/chocolate/service.py:7]: I am not good at py, why do we need () for it?

**Comment by gaocegege** [comment]
> By the way, I think we had better to support other common scales (like log-uniform scale) in feasible space.

I think so, would you mind opening an issue for it? @sperlingxx

File: pkg/suggestion/v1beta1/chocolate/service.py (hunk 1 above)
Comment (andreyvelich) [reply pkg/suggestion/v1beta1/chocolate/service.py]: I think we can remove it from here, it uses when we need to use more than 1 line for importing

**Comment by andreyvelich** [comment]
> /lgtm
> 
> By the way, I think we had better to support other common scales (like log-uniform scale) in feasible space.

In skopt we use log-uniform distribution: https://github.com/kubeflow/katib/blob/master/pkg/suggestion/v1beta1/skopt/base_service.py#L44.
To define it, max and min DOUBLE value is not enough?

**Comment by sperlingxx** [comment]
> > /lgtm
> > By the way, I think we had better to support other common scales (like log-uniform scale) in feasible space.
> 
> In skopt we use log-uniform distribution: https://github.com/kubeflow/katib/blob/master/pkg/suggestion/v1beta1/skopt/base_service.py#L44.
> To define it, max and min DOUBLE value is not enough?

Other packages used by katib also support multiple distributions:
* hyperopt supports `loguniform`, `lognormal`, `normal`, `qnormal` ...
* optuna supports `loguniform` and `uniform`

So, maybe is it better to add an option to configure distribution of continuous search space in `FeasibleSpace` ? And we can validate whether the tuning algorithm supporting specific distribution in `ValidateAlgorithmSettings` method of suggestion service.

**Comment by andreyvelich** [comment]
I believe @elikatsis wanted to hold this PR until end of this week.
We will merge it after.
/cc @gaocegege @sperlingxx

**Comment by gaocegege** [comment]
Oh SGTM. Thanks for the hold in time.

**Comment by andreyvelich** [comment]
I think we can merge these changes for Grid suggestion.
WDYT @anencore94 @johnugeorge @gaocegege ?

**Comment by gaocegege** [comment]
SGTM!

**Review by anencore94 — COMMENTED** [review]
`call_validate` call_validate to util class looks better! Thanks :) @andreyvelich

**Comment by gaocegege** [comment]
LGTM

**Pull request merged**